File: genData.m
function [A,Y,X_true,s,IDX,SNR]=genData(A,SNRdes,N0)

%%
[K,N] = size(A);
IDX=randi(N,1,N0);
fs=1/200; % Sampling frequency
T=25; % T time samples
maxFreq=randi([10 15],1,N0);
X_true=zeros(N,T);

for s=1:N0
    Sx=0;
    while  length(find(Sx))<2
        Minidx=1;
        %maxFreq=randi([5 15]);
        while length(Minidx)<2 || Minidx(end)-Minidx(1)<2
            Sx = randn(1,T); %-create white noise
            [b,a] = butter(2, maxFreq(s)*fs*2); % - get band of interest
            Sx = filtfilt(b, a, Sx); % -apply filter
            Minidx=find(abs(Sx)<(max(abs(Sx))*0.05));
        end
        
        Sx(1:Minidx(1)) = 0;
        Sx(Minidx(end):end) = 0;
        
    end
    
    X_true(IDX(s),:)=Sx;
end
X_true(IDX,:)=X_true(IDX,:)./repmat(max(abs(X_true(IDX,:)),[],2),1,25);
s=(abs(X_true)>0);
%%

Y0=A*X_true;
stdnoise = std(reshape(Y0,K*T,1))*10^(-SNRdes/20);
noise = stdnoise*randn(K,T);
Y=Y0+noise;
SNR = 10*log10(mean(var(Y0))/mean(var(noise)));

File: genData.py
# -*- coding: utf-8 -*-
"""
Created on Thu Aug 30 10:52:29 2018

@author: sofha
"""

def genData(A,SNRdes,N0):
    import numpy as np
    from butter_fil import butter_bandpass_filter
    #K,N =np.ndarray.size(A);
    K,N =A.shape;
   
    IDX = np.random.randint(1,N,size=N0);
    print(IDX)
    fs = 200; # Sampling frequency
    T = 100; # T time samples
    maxFreq = np.random.randint(10, 15,N0);
    X_true = np.zeros((N,T));
    
    
    for s in range(N0):
        print(s)
        Sx = 0
        it=1
        #while len(np.nonzero(Sx))<2:
        Minidx=1
        #while (len(np.nonzero(Minidx))<2 or Minidx[end]-Minidx[0]<2) and it==1:
        Sx = np.random.randn(1,T+1000); # creates white noise
        #plt.figure(s)
        #plt.plot(np.transpose(Sx))
        #plt.show
        it=2
        Sx = butter_bandpass_filter(Sx, maxFreq[s]-1/2, maxFreq[s]+1/2, fs, order=6)
        #.plot(np.transpose(Sx))
        Sx=Sx[0,500:500+T]
                 
                #Minidx = find(abs(Sx)<(max(abs(Sx))*0.05));
            #Sx[0:Minidx[0]] = 0;
            #Sx[Minidx[end]:end] = 0; 
        X_true[IDX[s],:] = Sx;
    Y0 = np.dot(A,X_true);
    stdnoise = np.std(np.ravel(Y0))*10**(-SNRdes/20);
    noise = stdnoise*np.random.randn(K,T);
    Y=Y0+noise;
    SNR = 10*np.log10(np.mean(np.var(Y0))/np.mean(np.var(noise)));

    
    return Sx,A,IDX,X_true,SNR,Y